.github/workflows/ci_pr.yml

```diff
@@ -106,33 +37,69 @@ jobs:
           - python-version: "3.10"
           - python-version: "3.11"
 
-    name: "Python ${{ matrix.python-version }} Unit Tests"
-    runs-on: ubuntu-20.04
-
-    env:
-      NOSEOPTS: "--with-timer ${{ matrix.additional-nose-opts }}"
-
     steps:
-
-    - name: Checkout WALinuxAgent repo
+    - name: Checkout WALinuxAgent
       uses: actions/checkout@v3
-
+    #
+    # We either install Python and the test dependencies, or download a pre-built virtual environment, depending on the
+    # use_virtual_environment flag.
+    #
     - name: Setup Python ${{ matrix.python-version }}
+      if: (!matrix.use_virtual_environment)
       uses: actions/setup-python@v4
       with:
         python-version: ${{ matrix.python-version }}
       env:
         PIP_TRUSTED_HOST: ${{ matrix.pip_trusted_host }}
 
     - name: Install dependencies
+      if: (!matrix.use_virtual_environment)
       id: install-dependencies
       run: |
         sudo env "PATH=$PATH" python -m pip install --upgrade pip
         sudo env "PATH=$PATH" pip install -r requirements.txt
         sudo env "PATH=$PATH" pip install -r test-requirements.txt
         sudo env "PATH=$PATH" pip install --upgrade pylint
 
+    - name: Setup Python ${{ matrix.python-version }} Virtual Environment
+      if: matrix.use_virtual_environment
+      id: install-venv
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y curl bzip2 sudo
+        curl -sSf --retry 5 -o /tmp/python-${{ matrix.python-version }}.tar.bz2 https://dcrdata.blob.core.windows.net/python/python-${{ matrix.python-version }}.tar.bz2
+        sudo tar xjf /tmp/python-${{ matrix.python-version }}.tar.bz2 --directory /
+        #
+        # The virtual environments for 2.6 and 3.4 have dependencies on OpenSSL 1.0, which is not available beyond Ubuntu 16. We use this script to patch the environments.
```

Comment:
Interestingly, the venv for 2.7 does not require patching when running directly on the VM (it does if one uses a container). The venvs for 2.6 and 3.4 require patching either when running directly on the VM or within a container.

I'll dig into this later to see if can simplify the setup further.